### PR TITLE
MAINT: LinearVectorFunction.J is ndarray closes #11886

### LINKED ISSUE #11886: Array shape mismatch in scipy.optimize
Shape mismatch of some np.matrix class
The issue can be solved easily by inserting
f0 = np.array(f0)[0]
below the line
f0 = fun.f
in _constraints.py

#### Reproducing code example:

```
def opt(x):
    return x[0]**2+x[1]**2

import scipy.optimize
A = np.matrix(np.diag([1,1]))
lin_cons = scipy.optimize.LinearConstraint(A, -1, np.inf)
    
params = scipy.optimize.minimize(opt, 2*[1], constraints = lin_cons)
```

#### Error message:

```

  File "C:\Users\Rapha\Desktop\Sigma\scratchbook.py", line 18, in <module>
    params = scipy.optimize.minimize(opt, 2*[1], constraints = lin_cons)

  File "c:\users\rapha\appdata\local\programs\python\python37\lib\site-packages\scipy\optimize\_minimize.py", line 595, in minimize
    constraints = standardize_constraints(constraints, x0, meth)

  File "c:\users\rapha\appdata\local\programs\python\python37\lib\site-packages\scipy\optimize\_minimize.py", line 825, in standardize_constraints
    old_constraints = new_constraint_to_old(con, x0)

  File "c:\users\rapha\appdata\local\programs\python\python37\lib\site-packages\scipy\optimize\_constraints.py", line 355, in new_constraint_to_old
    pcon = PreparedConstraint(con, x0)

  File "c:\users\rapha\appdata\local\programs\python\python37\lib\site-packages\scipy\optimize\_constraints.py", line 247, in __init__
    if np.any(f0[mask] < lb[mask]) or np.any(f0[mask] > ub[mask]):

  File "c:\users\rapha\appdata\local\programs\python\python37\lib\site-packages\numpy\matrixlib\defmatrix.py", line 195, in __getitem__
    out = N.ndarray.__getitem__(self, index)

IndexError: boolean index did not match indexed array along dimension 0; dimension is 1 but corresponding boolean dimension is 2  ...
```

### PR DESCRIPTION
**Pull request by andyfaff**
@mdhaber.

Makes sure `LinearVectorFunction.J` is an `np.ndarray`. Full explanation is in #11886.

### CODE REVIEW

**Comment by mdhaber** [comment]
Thanks. Could we add a test to https://github.com/scipy/scipy/blob/master/scipy/optimize/tests/test_minimize_constrained.py:

```
def test_bug_11886:
    def opt(x):
        return x[0]**2+x[1]**2
    A = np.matrix(np.diag([1,1]))
    lin_cons = scipy.optimize.LinearConstraint(A, -1, np.inf)    
    res = scipy.optimize.minimize(opt, 2*[1], constraints = lin_cons)
    assert_array_equal(res.x, [0, 0])
```

**Comment by andyfaff** [comment]
I tried adding a test, but then a PendingDeprecationWarning is emitted (which turns to an error with the test suite) when creating the matrix. I added your test though.

File: scipy/optimize/tests/test_minimize_constrained.py [hunk 1]
```diff
@@ -665,3 +666,14 @@ def constraintlcoh(x, v):
...
+
+def test_bug_11886():
+    def opt(x):
+        return x[0]**2+x[1]**2
+
+    with np.testing.suppress_warnings() as sup:
+        sup.filter(PendingDeprecationWarning)
+        A = np.matrix(np.diag([1, 1]))
+    lin_cons = LinearConstraint(A, -1, np.inf)
+    minimize(opt, 2*[1], constraints = lin_cons)
```
Comment (mdhaber) [inline scipy/optimize/tests/test_minimize_constrained.py:679]:
```suggestion
    minimize(opt, 2*[1], constraints = lin_cons)  # just checking that there are no errors
```

**Comment by mdhaber** [comment]
When CI finishes, I'll accept my suggestion (just a comment explaining why the test doesn't appear to test anything) and merge.

**Comment by mdhaber** [comment]
`test_heequb` failure is not from this PR. Thanks @andyfaff.

**Pull request merged**